packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart

```diff
@@ -0,0 +1,342 @@
+          '$address:$remotePort');
+      return null;
+    }
+    // TODO: The square-bracket enclosure for using the IPv6 loopback didn't
+    // appear to work, but when assigning to the IPv4 loopback device, netstat
+    // shows that the local port is actually being used on the IPv6 loopback
+    // (::1). While this can be used for forwarding to the destination IPv6
+    // interface, it cannot be used to connect to a websocket.
+    final String formattedForwardingUrl =
+        '$localPort:$_ipv4Loopback:$remotePort';
+    final List<String> command = <String>['ssh'];
+    if (isIpV6) {
+      command.add('-6');
+    }
+    if (sshConfigPath != null) {
+      command.addAll(<String>['-F', sshConfigPath]);
+    }
+    final String targetAddress =
+        isIpV6 && interface.isNotEmpty ? '$address%$interface' : address;
+    command.addAll(<String>[
+      '-nNT',
+      '-L',
+      formattedForwardingUrl,
+      targetAddress,
+    ]);
+    _log.fine("_SshPortForwarder running '${command.join(' ')}'");
+    final Process process = await _processManager.start(command);
+    process.exitCode.then((int c) {
+      _log.fine("'${command.join(' ')}' exited with exit code $c");
+    });
+    _log.fine('Set up forwarding from $localPort to $address port $remotePort');
+    return new _SshPortForwarder._(address, remotePort, localPort, process,
+        interface, sshConfigPath, isIpV6);
+  }
+
+  /// Kills the SSH forwarding command, then to ensure no ports are forwarded,
+  /// runs the SSH 'cancel' command to shut down port forwarding completely.
+  @override
+  Future<Null> stop() async {
+    // Kill the original SSH process if it is still around.
+    _process.kill();
+    // Cancel the forwarding request. See [start] for commentary about why this
+    // uses the IPv4 loopback.
+    final String formattedForwardingUrl =
+        '$_localPort:$_ipv4Loopback:$_remotePort';
+    final List<String> command = <String>['ssh'];
+    final String targetAddress = _ipV6 && _interface.isNotEmpty
+        ? '$_remoteAddress%$_interface'
+        : _remoteAddress;
+    if (_sshConfigPath != null) {
+      command.addAll(<String>['-F', _sshConfigPath]);
+    }
+    command.addAll(<String>[
+      '-O',
+      'cancel',
+      '-L',
+      formattedForwardingUrl,
+      targetAddress,
+    ]);
+    _log.fine(
+        'Shutting down SSH forwarding with command: ${command.join(' ')}');
+    final ProcessResult result = await _processManager.run(command);
+    if (result.exitCode != 0) {
+      _log.warning(
+          'Command failed:\nstdout: ${result.stdout}\nstderr: ${result.stderr}');
+    }
+  }
+
+  /// Attempts to find an available port.
+  ///
+  /// Finds the port by binding a socket to zero on the loopback interface, and
+  /// then unbinding. If this action could not be run for whatever reason, will
+  /// log a warning and return zero.
+  static Future<int> _potentiallyAvailablePort() async {
```

Comment:
this is a race condition. we should allocate a port once, not twice hoping that no other process allocates it in the meantime.

Reply:
So after snooping around, while it is a race condition, it seems there's no way to locally forward a port through SSH without previous knowledge of said port.

Would it be reasonable to implement this without use of SSH?  I believe it should be possible, just without secure data transfer.

Reply:
Fixed with bound socket.